features/FEATURE_BLE/targets/TARGET_CORDIO_ODIN_W2/HCIDriver.cpp

```diff
@@ -0,0 +1,352 @@
+
+        if (service_pack_index < service_pack_size)
+            send_service_pack_command();
+        else if (opcode == HCID_VS_WRITE_BD_ADDR) {
+           /* send an HCI Reset command to start the sequence */
+            HciResetCmd();
+            service_pack_transfered = true;
+        } else {
+            /* send BT device hardware address write command */
+            send_hci_vs_cmd(HCID_VS_WRITE_BD_ADDR);
+            cmd_opcode_ack_expected = HCID_VS_WRITE_BD_ADDR;
+        }
+    }
+
+    void hci_read_resolving_list_size(void)
+    {
+        /* if LL Privacy is supported by Controller and included */
+        if ((hciCoreCb.leSupFeat & HCI_LE_SUP_FEAT_PRIVACY) &&
+            (hciLeSupFeatCfg & HCI_LE_SUP_FEAT_PRIVACY)) {
+            /* send next command in sequence */
+            HciLeReadResolvingListSize();
+        } else {
+            hciCoreCb.resListSize = 0;
+
+            /* send next command in sequence */
+            hci_read_max_data_len();
+        }
+    }
+
+    void hci_read_max_data_len(void)
+    {
+        /* if LE Data Packet Length Extensions is supported by Controller and included */
+        if ((hciCoreCb.leSupFeat & HCI_LE_SUP_FEAT_DATA_LEN_EXT) &&
+            (hciLeSupFeatCfg & HCI_LE_SUP_FEAT_DATA_LEN_EXT)) {
+            /* send next command in sequence */
+            HciLeReadMaxDataLen();
+        } else {
+            /* send next command in sequence */
+            HciLeRandCmd();
+        }
+    }
+
+    DigitalOut      shutdown;                       // power/shutdown pin for bt device
+    DigitalOut      hci_rts;                        // request to sent pin
+    size_t          service_pack_index;             // Index of command to be recently sent over hci
+    bool            service_pack_transfered;        // Flag to notify if service pack is completely transferred or not
+    uint16_t        cmd_opcode_ack_expected;        // Command against which acknowledgment is expected
+    uint32_t        service_pack_size;              // size of service pack
+    char            *odin_service_pack ;            // Service pack needs to be provided by driver
+    vs_cmd_send_t   send_hci_vs_cmd ;               // callback function to call vendor specific call handler
+
+};
+
+} // namespace odin_w2
+} // namespace vendor
+} // namespace ble
+
+void ble::vendor::odin_w2::HCIDriver::do_initialize()
+{
+    cordio_callback_s    callback;
+
+    hci_rts = 1;            // Flow Control is OFF
+
+    shutdown = 0;           // BT Power is OFF
+    wait_ms(20);
+    shutdown = 1;           // BT Power is ON
+    wait_ms(500);
+
+    hci_rts = 0;            // Flow Control is ON
+
+    /* ODIN ble driver initialization function  */
+    cbCordio_Btinit(&callback);
+
+    odin_service_pack = callback.Service_pack;
+    send_hci_vs_cmd = callback.vs_command_callback;
+    service_pack_size = callback.service_pack_size;
+}
+
+void ble::vendor::odin_w2::HCIDriver::do_terminate()
+{
+    // TODO: ASRI
```

Comment:
left over todo, implement or remove and track in jira

Reply:
actually we have not found any method in Cordio Transport driver which should update HCI baudrate at run time if needed, it assumes baudrate being set at initialization never changes so it was just done in that perspective e.g if we really need to change baudrate first we would terminate current session of HCI and then update baudrate and re-open it but it is not required so I removed TODO in last commit, hope it have clear idea?